src/test_custom_request_schedule_manager.cc

```diff
@@ -0,0 +1,143 @@
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include <vector>
+
+#include "command_line_parser.h"
+#include "custom_request_schedule_manager.h"
+#include "doctest.h"
+#include "request_rate_worker.h"
+#include "test_load_manager_base.h"
+
+namespace triton::perfanalyzer {
+
+/// Class to test CustomRequestScheduleManager
+///
+class TestCustomRequestScheduleManager : public TestLoadManagerBase,
+                                         public CustomRequestScheduleManager {
+ public:
+  TestCustomRequestScheduleManager(
+      PerfAnalyzerParameters params, bool is_sequence_model = false,
+      bool is_decoupled_model = false, bool use_mock_infer = false)
+      : use_mock_infer_(use_mock_infer),
+        TestLoadManagerBase(params, is_sequence_model, is_decoupled_model),
+        CustomRequestScheduleManager(params, GetParser(), GetFactory()),
+        schedule_(params.schedule)
+  {
+  }
+
+  void TestSchedule(double request_rate, PerfAnalyzerParameters params)
+  {
+    int request_count = schedule_.size();
+    PauseWorkers();
+    ConfigureThreads(request_count);
+    GenerateSchedule(request_rate);
+
+    std::vector<std::chrono::nanoseconds> expected_timestamps;
+    std::chrono::nanoseconds timestamp(0);
+
+    for (float schedule_value : schedule_) {
+      float scaled_value = schedule_value / static_cast<float>(request_rate);
+
+      timestamp = std::chrono::duration_cast<std::chrono::nanoseconds>(
+          std::chrono::duration<float>(scaled_value));
+
+      expected_timestamps.push_back(timestamp);
+    }
+
+    size_t request_index = 0;
+    for (auto worker : workers_) {
+      auto timestamp = std::dynamic_pointer_cast<RequestRateWorker>(worker)
+                           ->GetNextTimestamp();
+      REQUIRE(timestamp.count() == expected_timestamps[request_index].count());
```

Comment:
Are the unit tests only testing that the correct number of timestamps are provided? Do the unit tests also need to make sure that the timestamps are the ones expected?

Reply:
As far as I understand, it's checking if the timestamps calculated in the test match the timestamps the workers are scheduled at. 
I have taken reference to this [check](https://github.com/triton-inference-server/perf_analyzer/blob/main/src/test_request_rate_manager.cc#L149) in **test_request_rate_manager.cc**
The .count() gives duration as an integer. 
Essentially, we are checking that the requests are scheduled at expected timestamps.

Reply:
I see, thanks for explaining. I believe you can check if the timestamps (std::chrono::nanoseconds) are equal directly and I don't believe the .count() adds anything to this check, but I'm okay with us leaving it if that's preferred. It's not worth slowing down this getting in.